### PR TITLE
fix(server, web): Prevent the user from setting a future date of birth

### PR DESCRIPTION
**Pull request by MohamedFBoussaid**
**Current Behaviour:** 

- If user enter birth date in the future it the age can be shown a negative value 
- Or if the image has wrong metadata. 

Screenshot 2023-12-17 at 21 54 14

**Suggestion:** 
Hide the age if the value is negative.

### CODE REVIEW

**Comment by martabal** [comment]
IMO, this is an incomplete solution. The ideal would be :

on the server side :
- have a sql constraint
- have a migration script to set to null all dates set to the future
- on the `person.service.ts`, don't allow future dates

on the client side:
- when choosing a date of birth, on the date picker, restrict the date to a past date

**Comment by MohamedFBoussaid** [comment]
> IMO, this is an incomplete solution. The ideal would be :
> 
> on the server side :
> 
> * have a sql constraint
> * have a migration script to set to null all dates set to the future
> * on the `person.service.ts`, don't allow future dates
> 
> on the client side:
> 
> * when choosing a date of birth, on the date picker, restrict the date to a past date

Yes, this could cover the case where the user will set a birth date in the future. 
But, today as I was using this feature, instead of setting a birthdate in 2002, by mistake have set it in 2022. While checking some picture taken in 2014, I was expecting to get 

> * person.service.ts

Hey @martabal, thanks for your comments. 
I have applied your feedbacks, is there any test environment that I can use to check the changes ?

**Comment by martabal** [comment]
> Yes, this could cover the case where the user will set a birth date in the future.
> But, today as I was using this feature, instead of setting a birthdate in 2002, by mistake have set it in 2022. While checking some picture taken in 2014, I was expecting to get

Indeed, it makes sense to keep your initial fix too

> Hey @martabal, thanks for your comments.
> I have applied your feedbacks, is there any test environment that I can use to check the changes ?

I don't know about this. I always test my changes locally on my machine.

File: server/src/infra/migrations/1702861544000-NullifyFutureBirthDates.ts
Comment (martabal) [inline server/src/infra/migrations/1702861544000-NullifyFutureBirthDates.ts:1]: Keep one migration file and put the update before the constraint

File: server/src/domain/person/person.service.ts
Comment (martabal) [inline server/src/domain/person/person.service.ts:205]: Can you add a unit test for this ? Should be in `person.service.spec.ts`

File: web/src/lib/components/faces-page/set-birth-date-modal.svelte
Comment (martabal) [inline web/src/lib/components/faces-page/set-birth-date-modal.svelte:59]: Can you add a `max` value in the date picker ?

File: server/src/domain/person/person.service.ts
Comment (MohamedFBoussaid) [reply server/src/domain/person/person.service.ts:205]: Thanks @martabal. The tests are added now.

File: web/src/lib/components/faces-page/set-birth-date-modal.svelte
Comment (MohamedFBoussaid) [reply web/src/lib/components/faces-page/set-birth-date-modal.svelte]: Thanks @martabal. The date picker will now have max value of today.

File: server/src/infra/migrations/1702861544000-NullifyFutureBirthDates.ts
Comment (MohamedFBoussaid) [reply server/src/infra/migrations/1702861544000-NullifyFutureBirthDates.ts:1]: Thanks @martabal. Updated.

**Review by martabal — APPROVED** [review]
Looks good to me, thank you for solving the issue

File: web/src/lib/components/faces-page/set-birth-date-modal.svelte
Comment (MohamedFBoussaid) [reply web/src/lib/components/faces-page/set-birth-date-modal.svelte:31]: The error message, was just to show to the user what went wrong. 
But I can update that to disable the button.

Comment (MohamedFBoussaid) [reply web/src/lib/components/faces-page/set-birth-date-modal.svelte:34]: @martabal I have updated the code, I suggest that we keep the error message, and disable the Set button in case of error. 
Also I have added the case where the user want to clear the birthdate. With my pervious changes, it will not be anymore possible once the birthdate is set.

File: server/src/infra/migrations/1702861544000-NullifyFutureBirthDatesAndAddCheckConstraint.ts
Comment (MohamedFBoussaid) [reply server/src/infra/migrations/1702861544000-NullifyFutureBirthDatesAndAddCheckConstraint.ts:8]: @martabal I have updated person.entity.ts, but I don't know if that is enough ?

**Comment by MohamedFBoussaid** [comment]
Sorry @martabal for the spam. 
Can you please double check my latest changes, before I merge (I am not sure about the @Check annotation)
Thanks.

File: web/src/lib/components/faces-page/set-birth-date-modal.svelte
Comment (martabal) [inline web/src/lib/components/faces-page/set-birth-date-modal.svelte:30]: Now that I think about it, do we need to check that the date is a future date if the input prevents it? If not, I don't think we need this `error`

**Review by martabal — COMMENTED** [review]
The check looks good to me, but it seems you have issues with typeorm migration script.

Don't forget to run `npm run format:fix` for the web and the server too

File: web/src/lib/components/faces-page/set-birth-date-modal.svelte
Comment (MohamedFBoussaid) [reply web/src/lib/components/faces-page/set-birth-date-modal.svelte:30]: The form will prevent the user from selecting a date in the future, as it will have max(date)=today. But This check is more if the user type it. 
I don't know how the component will behave if the user type 22.02.2030, will it just disable the set button ? or will it force the date to be today ? As I am not sure, I did all those checks. Better safe than sorry

**Comment by MohamedFBoussaid** [comment]
> The check looks good to me, but it seems you have issues with typeorm migration script.

I have tested the migration script on my db instance, and I didn't see any issues. What issues you think are there ?

> Don't forget to run `npm run format:fix` for the web and the server too
Done. Thanks.

**Comment by martabal** [comment]
> I have tested the migration script on my db instance, and I didn't see any issues. What issues you think are there ?

Typeorm is not happy https://github.com/immich-app/immich/actions/runs/7247973890/job/19742995638?pr=5803 :smile:

**Comment by MohamedFBoussaid** [comment]
> > I have tested the migration script on my db instance, and I didn't see any issues. What issues you think are there ?
> 
> Typeorm is not happy https://github.com/immich-app/immich/actions/runs/7247973890/job/19742995638?pr=5803 😄

Hahah I was just typing to let you know that I have found it lol. 
I was only seeing the top list of the Actions, I didn't thought there were more hahah. 
I will check it now

File: web/src/lib/components/faces-page/set-birth-date-modal.svelte
Comment (martabal) [reply web/src/lib/components/faces-page/set-birth-date-modal.svelte]: Don't worry for the error. In the worst scenario, the user will have a notification which explain what's the error.

This is the user flow with : 
```ts
export let birthDate: string | null;
```
```svelte
  const handleSubmit = () => {
    if (!birthDate) {
      return;
    }
    dispatch('updated', birthDate);
  };
```
```svelte
<Button type="submit" fullwidth disabled={!birthDate}>Set</Button>
```
https://github.com/immich-app/immich/assets/74269598/b130bb97-185f-47b5-a5f5-8cebcf5a08fd

Comment (MohamedFBoussaid) [reply web/src/lib/components/faces-page/set-birth-date-modal.svelte]: Ok I will do this changes, 
It make the code cleaner. 
But be aware of this corner case : 
If the user set a date of brith correct and save it, and for some reasons (s)he decide to delete the birth date of that person, I guess it will not be possible anymore as the birth date will be empty

Comment (martabal) [reply web/src/lib/components/faces-page/set-birth-date-modal.svelte]: Oh yes, I completely forgot that.
In this case, you just want to be sure that the date entered by the user is different from the original one.

**Comment by MohamedFBoussaid** [comment]
@martabal 
So I have fixed the issue with Typeorm, I was not doing the migration with the correct process. 
I was able to figure how to test this locally. 
Also I have made the logic simpler, and no need to have any error, bcs the date picker has a nice error message and clear. 
Also if the user can clear the date. 
And finally if the age is negative it will be hided. 
And the migration looks good

**Comment by martabal** [comment]
Yes, everything looks really good.

And again, a BIG thank you for your PR :smile:

**Review by martabal — APPROVED** [review]

**Comment by MohamedFBoussaid** [comment]
Thank you for your patience, while I am learning 🎉🎉

**Comment by MohamedFBoussaid** [comment]
Hey @martabal, is it possible to merge it. 
I don't have the write access to merge it. 
Thanks.

**Comment by martabal** [comment]
To be merged, it has to be approved by another member. It's the end of the year and some members are on vacation, just be patient :smile:

**Comment by MohamedFBoussaid** [comment]
> To be merged, it has to be approved by another member. It's the end of the year and some members are on vacation, just be patient 😄

Thanks @martabal, haha no I am not in a rush. 
I thought 1 approval was enough

File: web/src/lib/components/asset-viewer/detail-panel.svelte
Comment (MohamedFBoussaid) [reply web/src/lib/components/asset-viewer/detail-panel.svelte]: @alextran1502, nice catch, thanks. I don't remember deleting on purpose. Maybe I delete it while change Age code. 
I have put it back. Thanks

**Comment by martabal** [comment]
> Thanks @martabal, haha no I am not in a rush. I thought 1 approval was enough

Technically it was, but when a PR changes the server side, I prefer to have multiple approvals :sweat_smile:

**Comment by MohamedFBoussaid** [comment]
> > Thanks @martabal, haha no I am not in a rush. I thought 1 approval was enough
> 
> 
> 
> Technically it was, but when a PR changes the server side, I prefer to have multiple approvals :sweat_smile: 


You made the right decision 😅 https://github.com/immich-app/immich/pull/5803#discussion_r1431586806

**Pull request merged**
